homeassistant/components/balboa/sensor.py

```diff
@@ -0,0 +1,56 @@
+"""Support for Balboa sensors."""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+
+from pybalboa import SpaClient
+
+from homeassistant.components.sensor import SensorEntity
+from homeassistant.const import EntityCategory
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+from homeassistant.helpers.event import async_track_time_interval
+
+from . import BalboaConfigEntry
+from .entity import BalboaEntity
+
+REQUEST_FAULT_LOG_INTERVAL = timedelta(minutes=5)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: BalboaConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up the spa's sensors."""
+    spa = entry.runtime_data
+
+    async def request_fault_log(now: datetime | None = None) -> None:
+        """Request the most recent fault log."""
+        await spa.request_fault_log()
+
+    await request_fault_log()
+    entry.async_on_unload(
+        async_track_time_interval(hass, request_fault_log, REQUEST_FAULT_LOG_INTERVAL)
+    )
```

Comment:
If you register this in `async_added_to_hass` and dergister using `async_on_remove`, this won't be polled if someone disabled the entity

Reply:
Also, does this entity also work via the push method the other entities work with? If not, maybe it makes sense to just create this a polling entity

Reply:
> Also, does this entity also work via the push method the other entities work with? If not, maybe it makes sense to just create this a polling entity

Yes and no. The status updates come automatically from the spa once a connection is established. Fault log needs to be requested before the spa sends it back. It's not a direct response though, more of a, "when you get a chance, send me the latest fault log if you'd be so kind."

I could add a new event type callback in the library for fault log updates.

Reply:
Because if we get this message pushed, we could also consider creating it as an event entity

Reply:
Let's pretend we use an event entity, would I need to limit the firing of the event based on datetime?
As an example, let's pretend my spa has a fault of "Sensors are out of sync -- Call for service" from 5 days ago. If HA was rebooting or maybe I hadn't set up the integration yet, this still seems like a pretty important piece of information.

Reply:
Usually an event entity is just a single push. ![Screenshot_20250216-155227.png](https://github.com/user-attachments/assets/f3159049-00d0-4f3f-8505-001e951bbca8)

Currently they are a bit difficult to automate with, but this will be improved in the near future.

How quickly does the response happen? Usually event entities are used when something happens right now, and not when something happened 5 minutes ago to so say

Reply:
Yeah, that's what I was thinking with for event entities. After the request for a fault log, the spa pushes a message pretty quickly, but again, it will be the last known fault, which may be 5 days ago. I would argue that these are less events though. If my doorbell was pressed 5 days ago and HA missed it for whatever reason, I don't really care. If my spa had low water 5 days ago, I probably do still care.

Reply:
It also depends on what you receive, if you receive that same message every minute, I'm not sure if a timestamp is relevant, as the last_updated state would reflect that.

As in, I want HA to see that my doorbell rang 5 days ago, I don't want it to remind me that my doorbell rang.

Do we also detect if there are no faults?

Reply:
> Do we also detect if there are no faults?

To my knowledge, those who have worked on RE the spa communication protocol haven't found a way to clear faults. And there isn't a way to do so directly on my spa at least either. It's possible that a new spa doesn't have any faults yet, but also couldn't say for certain. They all seem to keep a 24 entry rolling log.

Reply:
> It also depends on what you receive, if you receive that same message every minute, I'm not sure if a timestamp is relevant, as the last_updated state would reflect that.

It would be the same message each time until the spa records a new one.

Reply:
Because then it maybe just makes sense to tell "hey, the water was empty 4 days ago". It's still as relevant as a couple of sensors.

But I'm now just pushing back and seeing what you can come up with, I don't have a definitive answer on this since I don't have the full context, while you have that, so don't see my comment as a hard requirement :)

Reply:
Oh I fully appreciate the discussion and an outsider's view ^_^